### PR TITLE
feat: introduce `types.ftl.go` to register sumtypes, which fixes encoding in unit tests

### PR DESCRIPTION
**Pull request by deniseli**
Fixes https://github.com/TBD54566975/ftl/issues/1577

This change introduces a new generated file `types.ftl.go` in the user-defined package itself to make sure sumtype registrations get called when running unit tests.

Notes:
* This change also includes a subpackage `subpackage` in the test data that is not yet unit tested. Unit tests should be added (identical to the parent package's unit tests) after https://github.com/TBD54566975/ftl/issues/1903 is resolved. Some work will also need to be done to extend the `types.ftl.go` generation to working on subpackages.
* This works by extending the existing logic operating on `mainModuleContext`, which means that code now generates not only just the literal `main` package, but also the conceptual _main_ package's additional artifacts. I couldn't come up with a name that would be better, so it's left as-is in this PR, but please let me know if you think of something better.

### CODE REVIEW

File: go-runtime/ftl/testdata/go/typeregistry/typeregistry_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+		Name:    "Object",
+		Input:   Object{S: "asdf"},
+		Encoded: `{"input":{"name":"Object","value":{"s":"asdf"}}}`,
+	},
+}
+
+func TestIngress(t *testing.T) {
+	ctx := ftltest.Context(ftltest.WithCallsAllowedWithinModule())
+
+	for _, test := range testCases {
```
Comment (alecthomas) [inline go-runtime/ftl/testdata/go/typeregistry/typeregistry_test.go:38]: Pull the list of structs into the test function so it's self-contained.

File: go-runtime/ftl/testdata/go/typeregistry/typeregistry_test.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
+package typeregistry
+
+import (
+	"ftl/builtin"
+	"testing"
+
+	"github.com/TBD54566975/ftl/go-runtime/encoding"
+	"github.com/TBD54566975/ftl/go-runtime/ftl"
+	"github.com/TBD54566975/ftl/go-runtime/ftl/ftltest"
+	"github.com/alecthomas/assert"
```
Comment (alecthomas) [inline go-runtime/ftl/testdata/go/typeregistry/typeregistry_test.go:10]: assert/v2 .... I do this all the time :(

**Review by alecthomas — APPROVED** [review]
Awesome. Have you played with it? What are the ergonomics like as you're updating the code/modifying the sum types?

File: go-runtime/ftl/testdata/go/typeregistry/typeregistry_test.go (hunk 2 above)
Comment (deniseli) [reply go-runtime/ftl/testdata/go/typeregistry/typeregistry_test.go:10]: ooo nice catch, thanks!

File: go-runtime/ftl/testdata/go/typeregistry/typeregistry_test.go (hunk 1 above)
Comment (deniseli) [reply go-runtime/ftl/testdata/go/typeregistry/typeregistry_test.go:38]: done

File: buildengine/discover_test.go [hunk 3]
```diff
@@ -38,7 +38,7 @@ func TestDiscoverModules(t *testing.T) {
 				DeployDir: "_ftl",
 				Schema:    "schema.pb",
 				Errors:    "errors.pb",
-				Watch:     []string{"**/*.go", "go.mod", "go.sum", "../../../go-runtime/ftl/**/*.go"},
+				Watch:     []string{"**/*.go", "go.mod", "go.sum", "../../../go-runtime/ftl/**/*.go", "../../../go-runtime/ftl/reflection/**/*.go"},
```
[the remark this answers is not in the document]
Comment (deniseli) [reply buildengine/discover_test.go]: Fixed

File: buildengine/discover_test.go [hunk 4]
```diff
@@ -156,7 +156,7 @@ func TestDiscoverModules(t *testing.T) {
...
```
[the remark this answers is not in the document]
Comment (deniseli) [reply buildengine/discover_test.go]: Fixed

**Comment by deniseli** [comment]
> Awesome. Have you played with it? What are the ergonomics like as you're updating the code/modifying the sum types?

It's been unintrusive, so good! Though I haven't tried it out in an IDE. Hopefully it works with everyone's workflows...

**Pull request merged**